spark/v3.3/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestMerge.java

```diff
@@ -214,11 +220,14 @@ public void testMergeIntoEmptyTargetInsertAllNonMatchingRows() {
             row(3, "emp-id-3") // new
             );
     assertEquals(
-        "Should have expected rows", expectedRows, sql("SELECT * FROM %s ORDER BY id", tableName));
+        "Should have expected rows",
+        expectedRows,
+        sql("SELECT * FROM %s ORDER BY id", selectTarget()));
   }
 
   @Test
   public void testMergeIntoEmptyTargetInsertOnlyMatchingRows() {
+    Assume.assumeTrue("Branch can only be used for non-empty table", branch == null);
```

Comment:
Looking into these test assumptions made me think more about what should be valid.

Currently, all of the test cases use a branch for all reads and writes, so there isn't anything testing implicit branching from the current table state. That is when writing, we will automatically create a branch from the current table state if it does not already exist. That was intended for convenience, so you can set an environment branch and auto-create on write.

For commands like MERGE that use the current table state, I'm wondering if that makes any sense because the command needs to create a read for the branch's state, which doesn't yet exist.

I'm leaning toward not allowing that for MERGE, UPDATE, or DELETE because it is so strange. Does that mean we shouldn't allow implicit branch creation for INSERT either? On one hand, that's what these tests rely on to write from the start into a branch. But on the other, there could be an expectation that writing to a branch that doesn't exist actually creates it from an empty state.

So...
1. I think we need tests for all of the operations that validate what happens when (a) the table is empty and an operation uses a branch that doesn't exist and (b) when the table is not empty and an operation uses a branch that doesn't exist
2. I think that MERGE, UPDATE, and DELETE should fail right away with a "Branch does not exist: %s" error (currently this is an NPE in job planning)
3. I think INSERT behavior should be clearly defined for branches, whether it is based on the current table state or an empty table state.

@aokolnychyi, @jackye1995, @danielcweeks, @amogh-jahagirdar, @namrathamyske, what do you think?

Reply:
Sorry had some errands and was not able to push everything during the weekend. 

Yes this is a point I also want to discuss. I used `Assume` to basically highlight this user experience explicit, glad we are having the same attentions to this issue here. I left the NPE there because this requires further discussion anyway. I can fix once we settle.

> when writing, we will automatically create a branch from the current table state if it does not already exist. That was intended for convenience, so you can set an environment branch and auto-create on write.

> I'm leaning toward not allowing that for MERGE, UPDATE, or DELETE because it is so strange

I also agree it's strange to create a new branch for operations like UPDATE and DELETE. However, if I think from the perspective of branching being an improved version of WAP, then it feels less strange because we are just saying the result of a write is staged in the specific branch until published.

The difference is that, in the WAP case, it won't explicitly say `DELETE FROM table.branch_xxx` and the branch is supposed to be hidden as an environment setting. So feels like this goes back to the argument about using identifier vs using an environment branch. If we consider them to be equivalent, then my logic would be: **because it makes sense to auto-create the branch for the WAP case, we need to match the experience for the identifier case even though it feels a bit awkward from SQL language perspective.**

This means we also support cases like `testMergeIntoEmptyTarget` for all branches. And extended further, following this logic, we will probably have to define **a non-existing branch to always be at the current table state**. So if user does `SELECT * from table.branch_not_exist` it will just read the current table instead of saying branch not found.

I think what Ryan is saying is following the opposite side of the logic that the WAP case and identifier case are not equivalent. My only concern would be it that would reduce a lot of the usability of branch, especially in the current situation that we consider the environment setting based approach to be difficult to achieve. At the same time, I am not sure if we really gain more consistency and clarity in this approach because there are still a lot of rules people need to understand and they might just get equally confused.

If we are saying special identifiers like `table.branch_`, `table.snapshot_id_`, etc. are hacks to work around query engine limitations, then it feels to me that we can accept some awkwardness to make it work for more use cases.

Would really like to know what others think.

Reply:
>  we will probably have to define a non-existing branch to always be at the current table state. So if user does `SELECT * from table.branch_not_exist` it will just read the current table instead of saying branch not found.

Actually, this is not needed. This can be explained as an experience similar to CTAS, where you can INSERT to a table before you create the table, the table is assumed to be there held in memory before committed. Here when running a MERGE/DELETE/UPDATE, we are basically saying the branch will be created before any scan and write, but it is only committed together with the write commit.

Reply:
> I also agree it's strange to create a new branch for operations like UPDATE and DELETE. However, if I think from the perspective of branching being an improved version of WAP, then it feels less strange because we are just saying the result of a write is staged in the specific branch until published.

I think we do want to use this for an improved version of WAP. But do we necessarily need to have implicit branch creation everywhere? I don't think that's required, and if we decide to add it we can always do it later.

> The difference is that, in the WAP case, it won't explicitly say DELETE FROM table.branch_xxx and the branch is supposed to be hidden as an environment setting

That's how WAP currently works, but we don't necessarily need branches to behave that way. We could create a `spark.wap.branch` setting that adds implicit branching and also ensures that all reads use that branch if it exists. That's well-defined behavior for an environment setting.

I'm more skeptical of implicit branch creation when performing normal branch writes using the `table.branch_x` naming. If you're specifically writing to a branch by name then it seems like it should exist before that point. If we require it to exist, then we avoid needing a behavior decision about how to initialize the state of an implicitly created branch.

> So if user does SELECT * from table.branch_not_exist it will just read the current table instead of saying branch not found.

I see your logic and I think that this follows, but I also don't think it is a good idea to define the state of a non-existent branch to be the current table state. Again, I think we could introduce an environment state that makes this the behavior specifically for WAP, but that shouldn't be the default for all branches and doesn't need to be added right now.

The most natural consequence of reading a branch that doesn't exist is to get an error. Otherwise, we get confusing behavior. For example, if you tell me you'll create a branch for me to look at in a table but I look at it before you prepare the data, I still get results. It also opens up problems when a name is slightly off; say I want to read _tag_ `q4_22_audit` but I actually read `q4_2022_audit` and get the current table state because the branch/tag doesn't exist. That would be really confusing for users.

I'm all for having ways to make WAP better using branches, but when we follow this line of reasoning it leads to confusing behavior in the more common cases. So I think we should make branches work in simple cases and then add more complex features on top.

I don't think this severely limits the utility of branches, since we can go through and add WAP on top as a next step.